### DISCUSSION TITLE
MQTT comando mosquitto_sub non trovato

### DISCUSSION THREAD

**Discussion by federoco** [indomus/forum#2345] [HUB: Home Assistant]
Ciao a tutti, 
ho collegato al mio impianto elettrico uno shelly 2.5 per controllare una tapparella.
Lo switch funziona perfettamente. 
Prima di integrarlo in HA con questa vostra [guida](https://indomus.it/guide/integrare-shelly-2-5-a-home-assistant-via-mqtt/), volevo testare lo shelly via MQTT.
Ho installato Eclipse Mosquitto come container Docker sul medesimo Raspberry dove gira HA, anche lui tramite Docker.
Il problema è che quando mi collego via SSH al Raspberry

``` 
mosquitto_sub -v -h 192.168.178.5 -u MIO_USER -P MIA_PASSWORD -t "Avvolgibile-Studio"
```

ricevo questo errore:
_-bash: mosquitto_sub: command not found_

Sto seguendo le indicazioni che fornite [QUI](https://indomus.it/formazione/mosquitto-mqtt-broker-comandi-utili/).
Ho pensato che, forse, nel container non erano installati i Mosquito Client tools, così li ho installati  tramite

```
apt-get install mosquitto-clients
```

Fatto questo non ricevo più il messaggio di errore "not found" ma il comando non va a buon fine.
Rimane il cursore lampeggiante e non accade nulla.

Qualche anima pia può darmi un suggerimento?
Grazie

**Comment by snakuzzo** [comment]
Non accade nulla perché probabilmente non passa alcun topic "Avvolgibile-Studio"
Prova così...

```
mosquitto_sub -v -t +/# -h 192.168.175.5 -u MIO_USER -P MIA_PASSWORD
```

**Reply by federoco** [reply]
Ciao Snakuzzo, grazie del suggerimento. L'ho provato appena tornato a casa ma purtroppo niente da fare.
Se provo a dare il tuo comando ricevo il solito messaggio:

```
pi@lampone:~ $ mosquitto_sub -v -t +/# -h 192.168.175.5 -u MIO_USER -P MIA_PASSWORD
-bash: mosquitto_sub: command not found
```

Se, invece, provo 

```
docker exec -it mosquitto sh  mosquitto_sub -v -t +/# -h 192.168.175.5 -u MIO_USER -P MIA_PASSWORD
```

come suggerito da Scialla, il messaggio di errore è il seguente:

```
sh: can't open 'mosquitto_sub': No such file or directory
```

Dalla verifica tramite il comando netstat -at mi pare che MQTT sia correttamente configurato e in ascolto.

```
pi@lampone:~ $ netstat -at
Active Internet connections (servers and established)
Proto Recv-Q Send-Q Local Address           Foreign Address         State      
tcp        0      0 0.0.0.0:http            0.0.0.0:*               LISTEN     
tcp        0      0 0.0.0.0:ssh             0.0.0.0:*               LISTEN     
tcp        0      0 0.0.0.0:https           0.0.0.0:*               LISTEN     
tcp        0      0 0.0.0.0:7020            0.0.0.0:*               LISTEN     
tcp        0      0 0.0.0.0:9000            0.0.0.0:*               LISTEN     
tcp        0      0 localhost:ipp           0.0.0.0:*               LISTEN     
tcp        0      0 0.0.0.0:1883            0.0.0.0:*               LISTEN     
tcp        0      0 0.0.0.0:5900            0.0.0.0:*               LISTEN     
tcp        0      0 Lampone.fritz.box:40000 0.0.0.0:*               LISTEN     
tcp        0      0 0.0.0.0:8123            0.0.0.0:*               LISTEN     
tcp        0      0 0.0.0.0:http-alt        0.0.0.0:*               LISTEN     
tcp        0    172 Lampone.fritz.box:ssh   Gipsy.fritz.box:49566   ESTABLISHED
tcp        0      0 localhost:42278         localhost:33201         ESTABLISHED
tcp        0      0 localhost:33201         localhost:42278         ESTABLISHED
tcp        0      0 172.18.0.1:52776        172.18.0.2:1883         ESTABLISHED
tcp        0      0 Lampone.fritz.box:38549 Lampone.fritz.box:1883  ESTABLISHED
tcp        0      0 Lampone.fritz.box:8123  192.168.178.100:57462   ESTABLISHED
tcp6       0      0 [::]:http               [::]:*                  LISTEN     
tcp6       0      0 [::]:ssh                [::]:*                  LISTEN     
tcp6       0      0 fe80::f54d:aa53:a:40000 [::]:*                  LISTEN     
tcp6       0      0 [::]:https              [::]:*                  LISTEN     
tcp6       0      0 localhost:ipp           [::]:*                  LISTEN     
tcp6       0      0 [::]:9000               [::]:*                  LISTEN     
tcp6       0      0 [::]:1883               [::]:*                  LISTEN     
tcp6       0      0 [::]:5900               [::]:*                  LISTEN     
tcp6       0      0 [::]:8123               [::]:*                  LISTEN     
tcp6       0      0 localhost:46578         localhost:1883          ESTABLISHED
tcp6       0      0 localhost:1883          localhost:46578         ESTABLISHED
```

Ho letto e riletto le vostre guide, la documentazione di Docker e di MQTT ma non riesco proprio a capire dove sbaglio.
Provo a fornire una panoramica della mia configurazione, se aiuta a definire il problema:
_Raspberry Pi a nome "Lampone" che risponde a 192.168.175.5 con sistema operativo Raspberry PI OS.
Qui sopra ho installato HA Core e il broker Eclipse Mosquitto, entrambi come contenitori Docker, tramite l'utility Docker compose.
Sempre tramire Docker sono installati anche zigbee2mqtt, broadlinkmanager, Caddy.
Modem in uso Fritzbox7530_
Attualmente su HA sono collegare varie lampade Ikea Tradfri via zigbee, dei Broadlink Universal Remote, il termostato della Vimar.
Tutto funziona perfettamente, anche da remoto.

**Reply by Scialla** [reply]
`docker exec -it mosquitto sh` dalla cartella dove hai il compose
ti apparirà il cursore col #
e dai:
`mosquitto_sub -v -t +/# -h localhost` oppure il comando che usavi prima
per interrompere ctrl+c
per uscire `exit`

**Reply by snakuzzo** [reply]
Allora...
ricominciamo da capo.
Se sull'host vuoi lanciare il comando `mosquitto_sub` devi installare i `mosquitto-clients` e lo fai con il comando...

```
sudo apt install mosquitto-clients
```

Se non vuoi installare il pacchetto sull'host, puoi usare il comando  `mosquitto_sub` presente nel container che hai creato con docker-compose, senza aprire la shell del container...

Puoi fare in uno di questi modi

- dalla directory in cui hai il file `docker-compose.yaml`...

es.
```
docker-compose run --rm mosquitto mosquitto_sub
```

dove `mosquitto` è il nome del service definito nel file `docker-compose.yaml`

- da qualsiasi directory specificando con il flag `-f` il path assoluto del file `docker-compose.yaml`

es.
```
docker-compose run -f /path/to/my/docker-compose.yaml --rm mosquitto mosquitto_sub
```

dove `mosquitto` è il nome del service definito nel file `docker-compose.yaml`

- oppure usando semplicemente il comando docker (da qualsiasi path)

es.
```
docker run --rm eclipse-mosquitto mosquitto_sub
```

dove `eclipse-mosquitto` è il nome dell'immagine di eclipse mosquitto 

Di seguito la prova di quello che ti ho appena scritto

![image](https://user-images.githubusercontent.com/7205371/219453852-e9968159-cd82-4f8c-b6ab-4ef09d5b732d.png)

**Reply by federoco** [reply]
Ciao snakuzzo, ciao Scialla,
grazie per il prezioso aiuto. Il paziente è vivo!
Ho attinto da entrambi i vostri suggerimenti e sono riuscito a collegarmi al topic.

Riporto il comando che ho usato, se può essere utile ad altri con problematiche simile:

```
docker compose run --rm mosquitto mosquitto_sub -v -h 192.168.178.5 -u MIO_USER -P MIA_PASSWORD -t "#"
```

che mostra tutti i topic di tutti i dispositivi.

Ho notato che la console ha mostrato quasi istantaneamente i topic in formato json delle lampade Ikea Tradfri collegate tramite zigbee2mqtt; mentre il topic dello shelly esce dopo 20, 30 secondi buoni. Come mai?
E perchè il formato è divero?
Questo, ad esempio, è la parte relativa ad una lampada Ikea:

```
homeassistant/light/MIO_ID/light/config {"availability":[{"topic":"zigbee2mqtt/bridge/state"}],"brightness":true,"brightness_scale":254,"color_mode":true,"command_topic":"zigbee2mqtt/ingresso1/set","device":{"identifiers":["zigbee2mqtt_MIO_ID"],"manufacturer":"IKEA","model":"TRADFRI LED bulb GU10 345/380 lumen, dimmable, white spectrum (LED2005R5)","name":"ingresso1","sw_version":"1.0.012"},"effect":true,"effect_list":["blink","breathe","okay","channel_change","finish_effect","stop_effect"],"json_attributes_topic":"zigbee2mqtt/ingresso1","max_mireds":454,"min_mireds":250,"name":"ingresso1","schema":"json","state_topic":"zigbee2mqtt/ingresso1","supported_color_modes":["color_temp"],"unique_id":"MIO_ID_light_zigbee2mqtt"}

```

e questo lo shelly

```
shellies/Avvolgibile-Studio/roller/0 stop
shellies/Avvolgibile-Studio/roller/0/pos 20
shellies/Avvolgibile-Studio/roller/0/stop_reason normal
shellies/Avvolgibile-Studio/roller/0/power 0.00
shellies/Avvolgibile-Studio/relay/power 0.00
shellies/Avvolgibile-Studio/roller/0/energy 0
shellies/Avvolgibile-Studio/relay/energy 0
shellies/Avvolgibile-Studio/input/1 0
shellies/Avvolgibile-Studio/input/0 0
shellies/Avvolgibile-Studio/temperature 33.14
shellies/Avvolgibile-Studio/temperature_f 91.65
shellies/Avvolgibile-Studio/overtemperature 0
shellies/Avvolgibile-Studio/temperature_status Normal
shellies/Avvolgibile-Studio/voltage 231.71
```

è normale che sia del tutto diverso?

Continuo a studiare...
Grazie ancora per il vostro tempo. Sto imparando molto.

**Reply by snakuzzo** [reply]
Quelli che visualizzi istantaneamente appena lanci `mosquitto_sub` sono quelli in retain
